### PR TITLE
splitting methods for transient Navier-Stokes solving

### PR DESCRIPTION
**Pull request by Huangzizhou**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -89,9 +89,9 @@ option(POLYFEM_SANITIZE_THREAD    "Sanitize Thread"        OFF)
 option(POLYFEM_SANITIZE_UNDEFINED "Sanitize Undefined"     OFF)
 
 #Solver
-option(POLYSOLVE_WITH_CHOLMOD          "Enable Cholmod library"            ON)
-option(POLYSOLVE_WITH_UMFPACK          "Enable UmfPack library"            ON)
-option(POLYSOLVE_WITH_SUPERLU          "Enable SuperLU library"            ON)
+option(POLYSOLVE_WITH_CHOLMOD          "Enable Cholmod library"            OFF)
```
Comment (teseoch) [inline CMakeLists.txt:92]: revert these changes, it is a bad idea to change the default

File: src/State.cpp [hunk 2]
```diff
@@ -483,13 +483,14 @@ namespace polyfem
...
+		extract_boundary_mesh_pressure();
 		// const std::string export_surface = args["export"]["surface"];
 		// if (!export_surface.empty())
-		extract_vis_boundary_mesh();
+			extract_vis_boundary_mesh();
 		logger().info("Done!");
 
 		const int prev_b_size = local_boundary.size();
-		problem->setup_bc(*mesh, bases, local_boundary, boundary_nodes, local_neumann_boundary);
+		problem->setup_bc(*mesh, bases, pressure_bases, local_boundary, boundary_nodes, local_neumann_boundary, pressure_boundary_nodes);
```
Comment (teseoch) [inline src/State.cpp:493]: does this affects FEM

File: src/assembler/utils/AssemblerUtils.cpp [hunk 3]
```diff
@@ -306,7 +307,7 @@ namespace polyfem
 		//else if(assembler == "Ogden")
 		//	ogden_elasticity_.local_assembler().compute_stress_tensor(bs, gbs, local_pts, fun, result);
 
-		else if (assembler == "Stokes") //WARNING stokes and NS dont have el_id
```
Comment (teseoch) [inline src/assembler/utils/AssemblerUtils.cpp:309]: is the warning solved? if not plz put back comment :)

File: src/mesh/Mesh.hpp [hunk 4]
```diff
@@ -113,6 +113,7 @@ namespace polyfem
 
 		//Nodal access
 		virtual RowVectorNd point(const int global_index) const = 0;
+		virtual int cell_vertex_(const int f_id, const int lv_id) const = 0;
```
Comment (teseoch) [inline src/mesh/Mesh.hpp:116]: why `_` in `cell_vertex_`?

File: src/mesh3D/Mesh3D.hpp [hunk 5]
```diff
@@ -29,13 +29,16 @@ namespace polyfem
...
-		inline int cell_edge(const int c_id, const int le_id) const { return mesh_.elements[c_id].es[le_id]; }
-		inline int face_vertex(const int f_id, const int lv_id) const { return mesh_.faces[f_id].vs[lv_id]; }
+		inline int n_face_vertices(const int f_id) const {return mesh_.faces[f_id].vs.size(); }
+		inline int n_cell_vertices(const int c_id) const {return mesh_.elements[c_id].vs.size(); }
+		inline int n_cell_faces(const int c_id) const {return mesh_.elements[c_id].fs.size(); }
+		inline int cell_vertex(const int c_id, const int lv_id) const {return mesh_.elements[c_id].vs[lv_id]; }
+		virtual int cell_vertex_(const int f_id, const int lv_id) const override { return mesh_.elements[f_id].vs[lv_id]; }
+		inline int cell_face(const int c_id, const int lf_id) const {return mesh_.elements[c_id].fs[lf_id]; }
+		inline int cell_edge(const int c_id, const int le_id) const {return mesh_.elements[c_id].es[le_id]; }
+		inline int face_vertex(const int f_id, const int lv_id) const {return mesh_.faces[f_id].vs[lv_id]; }
```
Comment (teseoch) [inline src/mesh3D/Mesh3D.hpp:39]: what changed here? indentation? and `cell_vertex_`?

File: src/problem/Problem.cpp [hunk 6]
```diff
@@ -21,15 +21,17 @@ namespace polyfem
 	{
 	}
 
-	void Problem::setup_bc(const Mesh &mesh, const std::vector<ElementBases> &bases, std::vector<LocalBoundary> &local_boundary, std::vector<int> &boundary_nodes, std::vector<LocalBoundary> &local_neumann_boundary)
+	void Problem::setup_bc(const Mesh &mesh, const std::vector<ElementBases> &bases, const std::vector<ElementBases> &pressure_bases, std::vector<LocalBoundary> &local_boundary, std::vector<int> &boundary_nodes, std::vector<LocalBoundary> &local_neumann_boundary, std::vector<int> &pressure_boundary_nodes)
```
Comment (teseoch) [inline src/problem/Problem.cpp:24]: this affects only splitting right?

File: src/state/StateInit.cpp [hunk 7]
```diff
@@ -116,6 +116,7 @@ namespace polyfem
 			{"t0", 0},
 			{"tend", 1},
 			{"time_steps", 10},
+			{"skip_frame", 1},
```
Comment (teseoch) [inline src/state/StateInit.cpp:119]: what it this

File: src/state/StateOutput.cpp [hunk 8]
```diff
@@ -518,6 +518,154 @@ namespace polyfem
 		return fs::weakly_canonical(fs::path(output_dir) / path).string();
 	}
 
+    void State::extract_boundary_mesh_pressure()
```
Comment (teseoch) [inline src/state/StateOutput.cpp:521]: is this the same as the other one? can it be refactored into 1 function with different params?

File: src/state/StateInit.cpp (hunk 7 above)
Comment (teseoch) [reply src/state/StateInit.cpp:119]: Ok got it, can you add to the other saves?

**Review by teseoch — CHANGES_REQUESTED** [review]

File: src/problem/Problem.cpp (hunk 6 above)
Comment (Huangzizhou) [reply src/problem/Problem.cpp:24]: Yes, this `pressure_boundary_nodes` is defined only to set up the pressure dirichlet bc for splitting. It's constructed in this function, same as what you do to construct the `boundary_nodes`.

File: src/State.cpp (hunk 2 above)
Comment (Huangzizhou) [reply src/State.cpp:493]: No, this `pressure_boundary_nodes` is defined only to set up the pressure dirichlet bc for splitting. It's constructed in this function, same as what you do to construct the `boundary_nodes`.

File: src/mesh3D/Mesh3D.hpp (hunk 5 above)
Comment (Huangzizhou) [reply src/mesh3D/Mesh3D.hpp:39]: I don't understand why they all changed either. In my only one commit in this file, only one line is highlighted.[https://github.com/Huangzizhou/polyfem/commit/70b7592f806ece5f551bc56b81bebac1e485006f#diff-49485c042b6ce08288115da8d5885345261e4fd4268e0c8c612f746f6258ab98](url)

File: src/state/StateInit.cpp (hunk 7 above)
Comment (Huangzizhou) [reply src/state/StateInit.cpp:119]: Sorry I don't understand

File: src/mesh3D/Mesh3D.hpp (hunk 5 above)
Comment (Huangzizhou) [reply src/mesh3D/Mesh3D.hpp:39]: I put a `_` at the end because there's a function which is not virtual and has the same name. Otherwise can I change the `cell_vertex` to virtual? (Sorry this is my first time using a virtual function)

File: src/mesh/Mesh.hpp (hunk 4 above)
Comment (Huangzizhou) [reply src/mesh/Mesh.hpp:116]: I put a `_` at the end because there's a function which is not virtual and has the same name in `Mesh3D` class. Otherwise can I change the original `cell_vertex` in `Mesh3D` to virtual? (Sorry this is my first time using a virtual function)

Comment (teseoch) [reply src/mesh/Mesh.hpp:116]: so remove the _ from `Mesh.hpp` and make `cell_vertex` in `Mesh3d` `override`, no need to make them virtual

File: src/mesh2D/Mesh2D.hpp [hunk 9]
```diff
@@ -70,6 +70,7 @@ namespace polyfem
 		void set_point(const int global_index, const RowVectorNd &p);
 
 		virtual RowVectorNd point(const int global_index) const override;
+		virtual int cell_vertex_(const int f_id, const int lv_id) const override { return mesh_.facets.vertex(f_id, lv_id); }
```
Comment (teseoch) [inline src/mesh2D/Mesh2D.hpp:73]: if this is the same as `face_vertex`, can you just call the other function?

File: src/state/StateInit.cpp (hunk 7 above)
Comment (teseoch) [reply src/state/StateInit.cpp:119]: add this `if (args["save_time_sequence"] && !(t % (int)args["skip_frame"]))` to the other time dependent solves :)

**Review by teseoch — CHANGES_REQUESTED** [review]

**Review by teseoch — APPROVED** [review]
